yucca/training/data_loading/YuccaDataset.py

```diff
@@ -116,6 +122,34 @@ def _transform(self, data_dict, case):
             data_dict = self.composed_transforms(data_dict)
         return self.to_torch(data_dict)
 
+    def replace_missing_modalities_with_zero_arrays(self, data):
+        assert data.dtype == "object", "allow missing modalities is true but dtype is not object"
```

Comment:
very related to the above comments

Reply:
My comment was poorly written. I think it would be cleaner if the unpack function didn't branch on the dtype, since this doesnt convey any information about why we are doing what we are doing to the reader of the code. What do you think about writting the code like this?

```python
supervised = self.task_type in ["classification", "segmentation"]

if self.allow_missing_modalities:
    image, label = self.unpack_with_zeros(data) if 
else:
    image, label = self.unpack(data)

def unpack_with_zeros(self, data, supervised: bool):
    assert data.dtype == "object", "allow missing modalities is true but dtype is not object"
    
    # First find the array with the largest array.
    # in classification this avoids setting the zero array to the 1d array with classes
    idx_largest_array = np.max([(idx, i.size) for idx, i in enumerate(data)], 0)[0]
    largest_shape = data[idx_largest_array].squeeze().shape
    
    # insert zeros
    for idx, i in enumerate(data):
        if np.isnan(data[idx]):
            data[idx] = np.zeros(largest_shape)
            
    if supervised:
        images = np.array([mod for mod in data[:-1]])
        label = data[-1:][0]
    else:
        images = np.array([mod for mod in data])
        label = None
        
    return images, label

def unpack(self, data, supervised: bool):
    if supervised:
           return data[:-1], data[-1:]
    return data, None
```


My thinking here is that it is clear from the context in `unpack_with_zeros` why we are doing what we are doing. And so is `unpack`. Also, i find that creating the zeros can naturally be part of the unpacking process, since this is exactly where we want them to appear.

Reply:
Also, because of the assert in the first line of `replace_missing_modalities_with_zero_arrays ` i guess `self.allow_missing_modalities` is actually a "force" missing modalities, since the dtype shouldn't be "object" unless we absolutely have to.

Reply:
good idea, I will do this instead (y)

Reply:
actually this was a very good suggestion, now that i've rewritten it it looks much much better